build/lib/src/generate/expected_outputs.dart

```diff
@@ -7,22 +7,115 @@ import '../builder/builder.dart';
 /// Collects the expected AssetIds created by [builder] when given [input] based
 /// on the extension configuration.
 Iterable<AssetId> expectedOutputs(Builder builder, AssetId input) {
-  var matchingExtensions =
-      builder.buildExtensions.keys.where((e) => input.path.endsWith(e));
-  return matchingExtensions
-      .expand((e) => _replaceExtensions(input, e, builder.buildExtensions[e]!));
+  return builder.parsedExtensions
+      .expand((extension) => extension.matchingOutputsFor(input));
 }
 
-Iterable<AssetId> _replaceExtensions(
-        AssetId assetId, String oldExtension, List<String> newExtensions) =>
-    newExtensions.map((n) => _replaceExtension(assetId, oldExtension, n));
-
-AssetId _replaceExtension(
-    AssetId assetId, String oldExtension, String newExtension) {
-  var path = assetId.path;
-  assert(path.endsWith(oldExtension));
-  return AssetId(
-      assetId.package,
-      path.replaceRange(
-          path.length - oldExtension.length, path.length, newExtension));
+// We can safely cache parsed build extensions for each builder since build
+// extensions are required to not change for a builder.
+final _parsedInputs = Expando<List<_ParsedBuildOutput>>();
+
+extension on Builder {
+  List<_ParsedBuildOutput> get parsedExtensions {
+    final existing = _parsedInputs[this];
+    if (existing == null) {
+      return _parsedInputs[this] = [
+        for (final entry in buildExtensions.entries)
+          _ParsedBuildOutput.parse(entry.key, entry.value),
+      ];
+    }
+
+    return existing;
+  }
+}
+
+class _ParsedBuildOutput {
+  static final RegExp _captureGroup = RegExp('{{}}');
+
+  final RegExp _pathMatcher;
+  final List<String> _outputs;
+  final bool _usesCaptureGroup;
+
+  _ParsedBuildOutput(this._pathMatcher, this._outputs, this._usesCaptureGroup);
+
+  factory _ParsedBuildOutput.parse(String input, List<String> outputs) {
+    final groups = _captureGroup.allMatches(input).iterator;
+
+    if (!groups.moveNext()) {
+      // The input does not contain a capture group, so we should simply match
+      // all files whose paths ends with the desired input.
+      return _ParsedBuildOutput(
+          RegExp('${RegExp.escape(input)}\$'), outputs, false);
+    }
+
+    final match = groups.current;
+
+    if (groups.moveNext()) {
+      throw ArgumentError(
+        'The builder declares an input "$input" which contains multiple '
+        'groups ("{{}}"). This is not supported.',
+      );
+    }
+
+    // When using a capture group in the build input, it must also be used in
+    // every output to ensure outputs have unique names.
+    for (final output in outputs) {
+      if (!_captureGroup.hasMatch(output)) {
```

Comment:
We want to assert exactly one match here also.

Reply:
Is this really necessary though? If I have `build_extensions: {'lib/{{}}.dart': ['lib/generated/{{}}/{{}}.g.dart']}` then I'd generate files in weird locations, but the generated files are still unique so it might not be problem? It's totally fair if we just want to avoid a footgun, I'm just wondering if there's something else I might be missing here.

Reply:
I think we just want to avoid the footgun here - we could always add support for that if there was a useful case that comes up, but we can't as easily remove the support.